## Data Collection

In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option(('display.max_columns'), (None))

In [2]:
#listings data obtained from http://insideairbnb.com/get-the-data
listings = "./data/source/listings.csv" #Listings data
listings_detailed = "./data/source/listings_detail.csv.gz" #Summary information and metrics for listings in New York City

df = pd.read_csv(listings, low_memory=False)
df_detailed = pd.read_csv(listings_detailed, low_memory=False)
df.shape, df_detailed.shape


((41533, 18), (41533, 75))

In [3]:
# Inspecting the data
df.head()

id                                      name  host_id  host_name  \
0  5136  Spacious Brooklyn Duplex, Patio + Garden     7378    Rebecca   
1  5203        Cozy Clean Guest Room - Family Apt     7490  MaryEllen   
2  5121                           BlissArtsSpace!     7356      Garon   
3  5178          Large Furnished Room Near B'way　     8967   Shunichi   
4  2595                     Skylit Midtown Castle     2845   Jennifer   

  neighbourhood_group       neighbourhood  latitude  longitude  \
0            Brooklyn         Sunset Park  40.66265  -73.99454   
1           Manhattan     Upper West Side  40.80380  -73.96751   
2            Brooklyn  Bedford-Stuyvesant  40.68535  -73.95512   
3           Manhattan             Midtown  40.76457  -73.98317   
4           Manhattan             Midtown  40.75356  -73.98559   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    275              21                  3  2022-08-10   
1     Private room     75               2                118  2017-07-21   
2     Private room     60              30                 50  2019-12-02   
3     Private room     68               2                559  2022-11-20   
4  Entire home/apt    175              30                 49  2022-06-21   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.03                               1               267   
1               0.73                               1                 0   
2               0.30                               2               322   
3               3.38                               1                79   
4               0.31                               3               365   

   number_of_reviews_ltm license  
0                      1     NaN  
1                      0     NaN  
2                      0     NaN  
3                     50     NaN  
4                      1     NaN

In [3]:
df_detailed.head()

id                        listing_url       scrape_id last_scraped  \
0  2595  https://www.airbnb.com/rooms/2595  20221204162430   2022-12-05   
1  5203  https://www.airbnb.com/rooms/5203  20221204162430   2022-12-05   
2  5136  https://www.airbnb.com/rooms/5136  20221204162430   2022-12-04   
3  5121  https://www.airbnb.com/rooms/5121  20221204162430   2022-12-05   
4  6848  https://www.airbnb.com/rooms/6848  20221204162430   2022-12-05   

            source                                      name  \
0      city scrape                     Skylit Midtown Castle   
1  previous scrape        Cozy Clean Guest Room - Family Apt   
2      city scrape  Spacious Brooklyn Duplex, Patio + Garden   
3      city scrape                           BlissArtsSpace!   
4      city scrape          Only 2 stops to Manhattan studio   

                                         description  \
0  Beautiful, spacious skylit studio in the heart...   
1  Our best guests are seeking a safe, clean, spa...   
2  We welcome you to stay in our lovely 2 br dupl...   
3  One room available for rent in a 2 bedroom apt...   
4  Comfortable studio apartment with super comfor...   

                               neighborhood_overview  \
0  Centrally located in the heart of Manhattan ju...   
1  Our neighborhood is full of restaurants and ca...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/f0813a11-40b2...     2845   
1  https://a0.muscache.com/pictures/103776/b37157...     7490   
2  https://a0.muscache.com/pictures/miso/Hosting-...     7378   
3  https://a0.muscache.com/pictures/2090980c-b68e...     7356   
4  https://a0.muscache.com/pictures/e4f031a7-f146...    15991   

                                  host_url      host_name  host_since  \
0   https://www.airbnb.com/users/show/2845       Jennifer  2008-09-09   
1   https://www.airbnb.com/users/show/7490      MaryEllen  2009-02-05   
2   https://www.airbnb.com/users/show/7378        Rebecca  2009-02-03   
3   https://www.airbnb.com/users/show/7356          Garon  2009-02-03   
4  https://www.airbnb.com/users/show/15991  Allen & Irina  2009-05-06   

  host_location                                         host_about  \
0  New York, NY  A New Yorker since (Phone number hidden by Air...   
1  New York, NY  Welcome to family life with my oldest two away...   
2  New York, NY  Rebecca is an artist/designer, and Henoch is i...   
3  New York, NY   I am an artist(painter, filmmaker) and curato...   
4  New York, NY  We love to travel. When we travel we like to s...   

  host_response_time host_response_rate host_acceptance_rate  \
0       within a day                72%                  22%   
1                NaN                NaN                  NaN   
2                NaN                NaN                  50%   
3       within a day                90%                  82%   
4     within an hour               100%                 100%   

  host_is_superhost                                 host_thumbnail_url  \
0                 f  https://a0.muscache.com/im/pictures/user/50fc5...   
1                 f  https://a0.muscache.com/im/users/7490/profile_...   
2                 f  https://a0.muscache.com/im/users/7378/profile_...   
3                 t  https://a0.muscache.com/im/pictures/user/72a61...   
4                 f  https://a0.muscache.com/im/users/15991/profile...   

                                    host_picture_url  host_neighbourhood  \
0  https://a0.muscache.com/im/pictures/user/50fc5...             Midtown   
1  https://a0.muscache.com/im/users/7490/profile_...     Upper West Side   
2  https://a0.muscache.com/im/users/7378/profile_...   Greenwood Heights   
3  https://a0.muscache.com/im/pictures/user/72a61...  Bedford-Stuyvesant   
4  https://a0.muscache.com/im/user

## Data Cleaning



Ill start by inspecting the nulls.

In [7]:
df.isnull().sum().sort_values(ascending=False)

license                           41532
reviews_per_month                  9393
last_review                        9393
name                                 13
host_name                             5
minimum_nights                        0
number_of_reviews_ltm                 0
availability_365                      0
calculated_host_listings_count        0
number_of_reviews                     0
id                                    0
room_type                             0
longitude                             0
latitude                              0
neighbourhood                         0
neighbourhood_group                   0
host_id                               0
price                                 0
dtype: int64

The licence information is irrelevant to the project and will be dropped.

There are 13 entries under the `name` field that are missing, as well as 9393 entries related to reviews. I will insepct these to see whats missing.

In [8]:
df = df.drop('license', axis=1)

In [9]:
df[df.name.isna()]

id name    host_id host_name neighbourhood_group  \
1827    2232600  NaN   11395220      Anna           Manhattan   
2831    4209595  NaN   20700823     Jesse           Manhattan   
2982    4370230  NaN   22686810   Michaël           Manhattan   
3099    4581788  NaN   21600904     Lucie            Brooklyn   
3296    4774658  NaN   24625694      Josh           Manhattan   
4395    6782407  NaN   31147528  Huei-Yin            Brooklyn   
5916    9325951  NaN   33377685  Jonathan           Manhattan   
6343    9787590  NaN   50448556    Miguel           Manhattan   
6675   10116081  NaN   51913270    Andrew           Manhattan   
6711   10052289  NaN   49522403   Vanessa            Brooklyn   
7816   12988898  NaN   71552588    Andrea               Bronx   
7839   12797684  NaN   69715276       Yan           Manhattan   
11742  21044649  NaN  151547086     Tanya               Bronx   

            neighbourhood  latitude  longitude        room_type  price  \
1827         East Village  40.73192  -73.98819  Entire home/apt    700   
2831    Greenwich Village  40.73323  -73.99294  Entire home/apt    225   
2982               Nolita  40.72100  -73.99536  Entire home/apt    215   
3099         Williamsburg  40.71370  -73.94378     Private room    150   
3296   Washington Heights  40.85111  -73.93009     Private room     40   
4395         Williamsburg  40.71354  -73.93882     Private room     45   
5916       Hell's Kitchen  40.76617  -73.98435  Entire home/apt    190   
6343               Harlem  40.80551  -73.95069  Entire home/apt    300   
6675              Midtown  40.75939  -73.96949  Entire home/apt    200   
6711          Brownsville  40.66409  -73.92314     Private room     80   
7816              Fordham  40.86046  -73.88378      Shared room    130   
7839      Upper West Side  40.79843  -73.96404     Private room    100   
11742             Norwood  40.87553  -73.88601     Private room     10   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
1827               60                 28  2015-06-08               0.27   
2831               30                  1  2015-01-01               0.01   
2982               30                  5  2016-01-02               0.05   
3099               30                  0         NaN                NaN   
3296               30                  0         NaN                NaN   
4395               30                  0         NaN                NaN   
5916               30                  1  2016-01-05               0.01   
6343               30                  0         NaN                NaN   
6675               30                  0         NaN                NaN   
6711               30                  3  2016-08-18               0.04   
7816               30                  0         NaN                NaN   
7839               30                  0         NaN                NaN   
11742               3                 24  2021-03-25               0.38   

       calculated_host_listings_count  availability_365  number_of_reviews_ltm  
1827                                1               359                      0  
2831                                1                 0                      0  
2982                                1                 0                      0  
3099                                1                 0                      0  
3296                                1                 0                      0  
4395                                1                 0                      0  
5916                                1                 0                      0  
6343                                5                 0                      0  
6675                                1                 0                      0  
6711                                1                 0                      0  
7816                                1                 0                      0  
7839                                2     

There doesn't appear to be any visual correlation between these listings in terms of location or host. Most of them have no reviews, and without a description highlighting amenities, we wont be able to include them in the comparisons. For now, I will name them 'Unnamed Listing' and we'll decide if any value can be extracted during the EDA step.

In [10]:
df.name.fillna('Unnamed Listing', inplace=True)

In [11]:
df.price.describe()

count    41533.000000
mean       221.978282
std        919.502236
min          0.000000
25%         80.000000
50%        131.000000
75%        220.000000
max      98159.000000
Name: price, dtype: float64

In [12]:
# The minimum value indicates there are listings for $0. 
df[df.price == 0]

id                                  name    host_id  \
20487  40560656  The Hoxton, Williamsburg - Cosy Room  273324213   
21001  41792753                 Mint House At 70 Pine  197053492   
21010  41740615            The James New York - NoMad  268417148   
21011  41740622       Gardens Suites Hotel by Affinia  269311462   
21148  42065543                  Broadway Plaza Hotel  307634016   
21152  42065545                       Carvi Hotel NYC  310429455   
21154  42065547                          Hotel Five44  308721299   
21155  42065555                          Pod Hotel BK  309714886   
21162  42065562                         Artezen Hotel  307633956   
21167  42065563                     Opera House Hotel  309772430   
21168  42065564                   The Wall Street Inn  314151200   
21224  42228997                  Sister City New York  314197504   
21278  42279171                        Leon Hotel LES  265458818   
21335  42384501                          Hotel Hayden  262458398   
21678  43035720     Executive Hotel LeSoleil New York  318559292   
21787  43078550                     The Lex Hotel NYC  334334264   
21788  43078552                             SIXTY LES  342053968   
21900  43205598                      Wellington Hotel  335389657   
22006  43247386         6 Columbus Central Park Hotel  335072254   
22010  43247472                              Hotel 57  324955773   
22014  43247631                       Belvedere Hotel  318788301   
22536  44567521            The Standard, East Village  360662584   
23282  46059074  CITIZENM NEW YORK TIMES SQUARE HOTEL  373324108   
23356  46087899                      The William Vale  373522899   
23465  46336133                Equinox Hotel New York  375044940   
23466  46251446        citizenM New York Bowery hotel  374516933   
23742  46723973                          Ink 48 Hotel  376877842   
24518  48089897                          Condor Hotel  261016212   
24785  48417136                          HGU New York  390810530   
24795  48325676                          Hotel Beacon  390077597   

                              host_name neighbourhood_group  \
20487           The Hoxton Williamsburg            Brooklyn   
21001             Mint House At 70 Pine           Manhattan   
21010                   The James NoMad           Manhattan   
21011   Gardens Suites Hotel By Affinia           Manhattan   
21148                    Broadway Plaza           Manhattan   
21152                       Carvi Hotel           Manhattan   
21154                      Hotel Five44           Manhattan   
21155                         Pod Hotel            Brooklyn   
21162                     Artezen Hotel           Manhattan   
21167                 Opera House Hotel               Bronx   
21168               The Wall Street Inn           Manhattan   
21224                       Sister City           Manhattan   
21278                    Leon Hotel LES           Manhattan   
21335                      Hotel Hayden           Manhattan   
21678                Executive LeSoleil           Manhattan   
21787                         Lex Hotel           Manhattan   
21788                         Sixty Les           Manhattan   
21900                  Wellington Hotel           Manhattan   
22006  Six Columbus  Central Park Hotel           Manhattan   
22010                          Hotel 57           Manhattan   
22014                              Beth           Manhattan   
22536                      The Standard           Manhattan   
23282                          CitizenM           Manhattan   
23356                  The William Vale            Brooklyn   
23465                     Equinox Hotel           Manhattan   
23466                          CitizenM           Manhattan   
23742                       Ink48 Hotel           Manhattan   
24518                  The Condor Hotel            Brooklyn   
24785                            Marcos           Manhattan   
24795         

It looks like there are 30 listings with a price of $0. This may be true or it may be an error, but there is no efficient way to check. Additionally, these entries are all listed as Hotel room for the `room_type`. Since the start-up will not be including these in their listing, we may drop them all together. For now, I will drop these entries since they wont help with the price.

In [13]:
df = df[df.price != 0]
df.price.describe()

count    41503.000000
mean       222.138737
std        919.815135
min         10.000000
25%         80.000000
50%        131.000000
75%        220.000000
max      98159.000000
Name: price, dtype: float64

I will replace the null values in the `reviews_per_month` column with 0. As for the `last_review` column, it may be of help later in measuring the activity of listings but it will be dropped for now.

In [14]:
df.reviews_per_month.fillna(0)
df = df.drop('last_review', axis=1)
df.columns

Index(['id', 'name', 'host_id', 'host_name', 'neighbourhood_group',
       'neighbourhood', 'latitude', 'longitude', 'room_type', 'price',
       'minimum_nights', 'number_of_reviews', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365',
       'number_of_reviews_ltm'],
      dtype='object')

`id` and `name` are the only identifiers needed. The `host_name` and `host_id` will be dropped.

In [15]:
df_final = df.drop(['host_name', 'host_id'], axis=1)
print(df_final.columns)

Index(['id', 'name', 'host_id', 'neighbourhood_group', 'neighbourhood',
       'latitude', 'longitude', 'room_type', 'price', 'minimum_nights',
       'number_of_reviews', 'reviews_per_month',
       'calculated_host_listings_count', 'availability_365',
       'number_of_reviews_ltm'],
      dtype='object')


In [ ]:
df_detailed.columns()

The features I plan to investigate effect on price are: location, property type, and number of rooms.

For location: I'll be using the `neighbourhood` and `neighbourhood_group` as classifiers. The geocode location can be added later for a more in depth analysis or new feature engineering.

The `room_type` entry comprises of the following unique values: 'Entire home/apt', 'Private room', 'Hotel room', 'Shared room'. Hotel room entries will be dropped in a future step.

`accommodates` and `bedrooms` are integers relating to number of guests.
`bathrooms_text` contains string data about the number of baths.
These will also be added in the EDA step.

`amenities` is a json object of self-reported available amenities. This may be used in a future step.

To investigate local market conditions: I will investigate the hosts mentioned `availability_365`. This may be used to calculate occupancy rate in a future step.

Past rental data in this case will include: the `number_of_reviews` and `number_of_reviews_ltm` which is specific to the past 12 months.

The remaining features will be left in during the EDA step to check for other correlations. 

In [16]:
listings_final = df_final.merge(df_detailed[['accommodates', 'bedrooms', 'bathrooms_text', 'amenities', 'id','beds']], on='id', how='left')

In [17]:
df_final.shape, listings_final.shape

((41503, 15), (41503, 20))

In [18]:
listings_final.to_csv('./data/listings_wrangled', index=False)